---
title: "AMR analysis"
author: "Lisa"
date: "12/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Import/install packages
```{r, warning=FALSE, message=FALSE}
packages <- c("magrittr","dplyr","ggplot2","stringr","cooccur","visNetwork","igraph","tidyr","ComplexHeatmap","phylogram","proxy")
for (package in packages){
  if (!require(package, character.only = TRUE)) install.packages(package)
  require(package, character.only = TRUE)
}
```

Import resfinder file
```{r}
resfinder <- read.delim("../results/bactofidia_output_all/stats/ResFinder.tsv")
```

Keep all valid samples
```{r}
all_strains <- read.delim("../results/all_ecoli_rgnosis.txt",header = FALSE)
resfinder %<>% filter(sub(".fna","",X.FILE) %in% all_strains$V1) %>% filter(X.FILE != "ECO-JSC-RGN-103823.fna")
```

Filter based on coverage threshold
```{r}
coverage_threshold <- 95
resfinder %<>% filter(X.COVERAGE >= coverage_threshold)
```

Add information on origin (plasmid/chromosome) of the contigs
```{r}
contig_classifications <- read.csv("../results/EC_combined_output.csv")
resfinder <- left_join(resfinder,contig_classifications %>% select(contig_name,classification),by = c("SEQUENCE"="contig_name"))
```

Get the MLST
```{r}
MLST <- read.delim("../results/bactofidia_output_all/stats/MLST.tsv",header = FALSE)
colnames(MLST) <- c("strain","species","ST","adk","fumC","gyrB","icd","mdh","purA","recA")
#It also contains some non-Ecoli strains that we are not interested in, let's filter them out
all_ecoli <- read.delim("../results/all_ecoli_rgnosis.txt",header = FALSE)
MLST %<>% filter(sub(".fna","",strain) %in% all_ecoli$V1)
```

Add ST to resfinder file, and group by ST131/non-ST131
```{r}
resfinder %<>% merge(MLST %>% select(strain,ST),by.x = "X.FILE", by.y = "strain")
resfinder %<>% mutate(ST131=ifelse(ST=="131","ST131","Other"))
#add sample that is ST131 but not recognised by MLST
resfinder %<>% mutate(ST131=ifelse(X.FILE=="ECO-JSC-MOD-104871.fna","ST131",ST131))
```

Count AMR genes per sample
```{r}
count_AMR <- resfinder %>% group_by(X.FILE) %>% summarise(count=n())
```

```{r}
ggplot(count_AMR,aes(x=count,fill=ST131))+
  geom_histogram(binwidth=1,stat = 'count')+
  xlim(0,NA)+
  #scale_x_continuous(breaks = seq(0,40,1))+
  labs(x="Nr. of AMR genes",y="Nr. of samples")
```
```{r}
ggplot(count_AMR,aes(y=count,x=factor(ST131,levels = c("ST131","Other")),fill=factor(ST131,levels = c("ST131","Other"))))+
  geom_boxplot()+
  theme(legend.position = "none")+
  stat_compare_means(method = "t.test", label.y=15)+
  labs(x = NULL, y = "Nr. of ARGs")
```
```{r}
t.test(count_AMR %>% filter(ST131=="ST131") %>% select(count),count_AMR %>% filter(ST131=="Other") %>% select(count))
```


Count samples per ARG
```{r}
count_samples <- resfinder %>% group_by(GENE,classification,ST131) %>% summarise(count=n())
count_samples %<>% group_by(GENE) %>% mutate(total_count=sum(count))
```

Only display ARGs present a least 10 times
```{r}
count_samples_selected <- count_samples %>% filter(total_count>=10)
```

```{r}
ggplot(count_samples_selected, aes(x=count,y=reorder(GENE,+total_count), fill=classification))+
  geom_bar(position= 'stack', stat = 'identity')+
  labs(y="ARG")+
  facet_wrap(~factor(ST131,levels = c("ST131","Other")), scales = "free_x")+
  labs(fill="Predicted location")+
  scale_fill_manual(values=c("#de6909","#218f61"))
```


## Co-occurrence analysis for ST131
Are certain AMR genes more frequently found together?

Extract ST131 samples
```{r}
ST131_strains <- read.delim("../results/sample_summaries_ST131.csv",sep = ';') 
resfinder_ST131 <- resfinder %>% filter(sub(".fna","",X.FILE) %in% ST131_strains$Sample)
```

Transform the resfinder data to get a presence absence matrix for each gene / sample
```{r}
cooccur_data <- resfinder_ST131 %>% select(X.FILE,GENE)
pres_abs <- as.data.frame.matrix(table(cooccur_data[2:1]))
```

Some samples contain multiple copies of the same AMR gene. For this analysis, we are just interested in whether a gene is present at all (1) or not (0), so presences > 1 we'll set to 1.
```{r}
pres_abs %<>% mutate_all(funs(ifelse(. > 1,1,.)))
```

Use cooccur package to perform co-occurrence analysis
```{r}
cooccurrence_ST131 <- cooccur(pres_abs, spp_names = TRUE)
cooccur_results <- cooccurrence_ST131$results
```
Because we test a lot of hypotheses at once, we should correct for multiple testing. We use the Benjamini Hochberg method which controls the false discovery rate (FDR) instead of the family-wise error rate.
```{r}
cooccur_results %<>% mutate(p_lt_adj=p.adjust(p_lt,method = "BH"),p_gt_adj=p.adjust(p_gt,method = "BH"))
```

Filter to keep only significant values
```{r}
cooccur_sign <- cooccur_results %>% filter(p_lt_adj <= 0.05 | p_gt_adj <= 0.05)
```

### Visualisation with visNetwork

Build nodes
```{r}
#Find the genes that have significant results (so that are either in s1 or s2 of cooccur_sign)
sign_genes <- (unique(c(cooccur_sign$sp1_name,cooccur_sign$sp2_name)))

#Calculate per gene how often it is found on plasmids
plasmid_fractions <- count_samples %>% select(-total_count) %>% spread(classification,count) %>% replace(is.na(.),0) %>% mutate(plasmid_fraction=plasmid/(chromosome+plasmid))

#Build node file with significant genes and their id's (row numbers)
nodes <- data.frame(
  label=rownames(pres_abs),
  id=1:nrow(pres_abs),
  value=rowSums(pres_abs)) %>% filter(label %in% sign_genes)

#Add plasmid fractions
nodes %<>% left_join(.,plasmid_fractions,by=c("label"="GENE"))

#Convert plasmid fraction to colour
values <- nodes$plasmid_fraction
ii <- cut(values, breaks = seq(0,1, len = 100), 
          include.lowest = TRUE)
nodes$color <- colorRampPalette(c("#de6909","#cfcfcf","#218f61"))(99)[ii]

```

Build edges
```{r}
edges <- data.frame(
  from=cooccur_sign$sp1,
  to=cooccur_sign$sp2,
  #Colour according to positive / negative interaction
  color=list(color=ifelse(cooccur_sign$p_lt <= 0.05,'#D5232F','#4CB942'),opacity=0.5),
  #The smaller the p value, the thicker the edge
  value=0.05-apply(cooccur_sign[,c('p_lt','p_gt')],1,FUN = min),
  length=1
  )
```

Plot the network

```{r}
visNetwork(nodes, edges) %>%
    visIgraphLayout(layout="layout_with_kk")
```

### Co-occurrence analysis for other E.coli
Are certain AMR genes more frequently found together?

Extract other E.coli samples
```{r}
resfinder_other <- resfinder %>% filter(!sub(".fna","",X.FILE) %in% ST131_strains$Sample)
```

Transform the resfinder data to get a presence absence matrix for each gene / sample
```{r}
cooccur_data <- resfinder_other %>% select(X.FILE,GENE)
pres_abs <- as.data.frame.matrix(table(cooccur_data[2:1]))
```

Some samples contain multiple copies of the same AMR gene. For this analysis, we are just interested in whether a gene is present at all (1) or not (0), so presences > 1 we'll set to 1.
```{r}
pres_abs %<>% mutate_all(funs(ifelse(. > 1,1,.)))
```

Use cooccur package to perform co-occurrence analysis
```{r}
cooccurrence_other<- cooccur(pres_abs, spp_names = TRUE)
cooccur_results <- cooccurrence_other$results
```
Because we test a lot of hypotheses at once, we should correct for multiple testing
```{r}
cooccur_results %<>% mutate(p_lt_adj=p.adjust(p_lt,method = "BH"),p_gt_adj=p.adjust(p_gt,method = "BH"))
```

Filter to keep only significant values
```{r}
cooccur_sign <- cooccur_results %>% filter(p_lt_adj <= 0.05 | p_gt_adj <= 0.05)
```

Or we can only focus only on the bla (beta lactamase) genes
```{r}
cooccur_bla <- cooccur_sign %>% filter(grepl('bla',sp1_name) | grepl('bla',sp2_name))
```

### Visualisation with visNetwork

Build nodes
```{r}
dataset <- cooccur_sign
#Find the genes that have significant results (so that are either in s1 or s2 of cooccur_sign)
sign_genes <- (unique(c(dataset$sp1_name,dataset$sp2_name)))

#Build node file with significant genes and their id's (row numbers)
nodes <- data.frame(
  label=rownames(pres_abs),
  id=1:nrow(pres_abs),
  value=rowSums(pres_abs)) %>% filter(label %in% sign_genes)

#Add plasmid fractions
nodes %<>% left_join(.,plasmid_fractions,by=c("label"="GENE"))

#Convert plasmid fraction to colour
values <- nodes$plasmid_fraction
ii <- cut(values, breaks = seq(0,1, len = 100), 
          include.lowest = TRUE)
nodes$color <- colorRampPalette(c("#de6909","#cfcfcf","#218f61"))(99)[ii]
```


Build edges
```{r}
edges <- data.frame(
  from=dataset$sp1,
  to=dataset$sp2,
  #Colour according to positive / negative interaction
  color=list(color=ifelse(dataset$p_lt <= 0.05,'#D5232F','#4CB942'),opacity=0.5),
  #The smaller the p value, the thicker the edge
  value=0.05-apply(dataset[,c('p_lt','p_gt')],1,FUN = min),
  length=1
  )
```

Plot the network
```{r}
visNetwork(nodes, edges) %>%
    visIgraphLayout(layout="layout_with_kk")
```
### And for the full dataset?

Transform the resfinder data to get a presence absence matrix for each gene / sample
```{r}
cooccur_data <- resfinder %>% select(X.FILE,GENE)
pres_abs <- as.data.frame.matrix(table(cooccur_data[2:1]))
```

Some samples contain multiple copies of the same AMR gene. For this analysis, we are just interested in whether a gene is present at all (1) or not (0), so presences > 1 we'll set to 1.
```{r}
pres_abs %<>% mutate_all(funs(ifelse(. > 1,1,.)))
```

Use cooccur package to perform co-occurrence analysis
```{r}
cooccurrence_all <- cooccur(pres_abs, spp_names = TRUE)
cooccur_results <- cooccurrence_all$results
```
Because we test a lot of hypotheses at once, we should correct for multiple testing
```{r}
cooccur_results %<>% mutate(p_lt_adj=p.adjust(p_lt,method = "BH"),p_gt_adj=p.adjust(p_gt,method = "BH"))
```

Filter to keep only significant values
```{r}
cooccur_sign <- cooccur_results %>% filter(p_lt_adj <= 0.05 | p_gt_adj <= 0.05)
```

Or we can only focus only on the bla (beta lactamase) genes
```{r}
cooccur_bla <- cooccur_sign %>% filter(grepl('bla',sp1_name) | grepl('bla',sp2_name))
```

### Visualisation with visNetwork

Build nodes
```{r}
dataset <- cooccur_sign
#Find the genes that have significant results (so that are either in s1 or s2 of cooccur_sign)
sign_genes <- (unique(c(dataset$sp1_name,dataset$sp2_name)))

#Build node file with significant genes and their id's (row numbers)
nodes <- data.frame(
  label=rownames(pres_abs),
  id=1:nrow(pres_abs),
  value=rowSums(pres_abs)) %>% filter(label %in% sign_genes)

#Add plasmid fractions
nodes %<>% left_join(.,plasmid_fractions,by=c("label"="GENE"))

#Convert plasmid fraction to colour
values <- nodes$plasmid_fraction
ii <- cut(values, breaks = seq(0,1, len = 100), 
          include.lowest = TRUE)
nodes$color <- colorRampPalette(c("#de6909","#cfcfcf","#218f61"))(99)[ii]
```


Build edges
```{r}
edges <- data.frame(
  from=dataset$sp1,
  to=dataset$sp2,
  #Colour according to positive / negative interaction
  color=list(color=ifelse(dataset$p_lt <= 0.05,'#D5232F','#4CB942'),opacity=0.5),
  #The smaller the p value, the thicker the edge
  value=0.05-apply(dataset[,c('p_lt','p_gt')],1,FUN = min),
  length=1
  )
```

Plot the network
```{r}
visNetwork(nodes, edges) %>%
    visIgraphLayout(layout="layout_with_kk")
```

### Heatmaps

I wanna try a heatmap for visualisation

```{r}
pres_abs_general <- as.data.frame.matrix(table(resfinder %>% select(GENE,X.FILE))) %>% mutate_all(funs(ifelse(. > 1,1,.)))
```

If we want to distinguish between genes on chromosome and genes on plasmid:
```{r}
#we'll code plasmid genes as a '1'
pres_abs_plasmid <- as.data.frame.matrix(table(resfinder %>% filter(classification=="plasmid") %>% select(GENE,X.FILE))) %>% mutate_all(funs(ifelse(.>=1,1,.)))
#we'll code chromosome genes as a  2'
pres_abs_chromo <- as.data.frame.matrix(table(resfinder %>% filter(classification=="chromosome") %>% select(GENE,X.FILE))) %>% mutate_all(funs(ifelse(.>=1,2,.)))
#but the number of rows and columns don't match yet!
columns2plasmid <- setdiff(colnames(pres_abs_general),colnames(pres_abs_plasmid))
for(i in columns2plasmid) {
    pres_abs_plasmid[,i] <- 0
}
rows2plasmid <- setdiff(rownames(pres_abs_general),rownames(pres_abs_plasmid))
for(i in rows2plasmid) {
    pres_abs_plasmid[i,] <- 0
}
rows2chromo <- setdiff(rownames(pres_abs_general),rownames(pres_abs_chromo))
for(i in rows2chromo) {
    pres_abs_chromo[i,] <- 0
}
#sort alphabetically
pres_abs_chromo <- pres_abs_chromo[order(row.names(pres_abs_chromo)), ]
pres_abs_plasmid <- pres_abs_plasmid[order(row.names(pres_abs_plasmid)), ]

#add the two matrices, if there is a '3' it has both
pres_abs_class <- pres_abs_plasmid + pres_abs_chromo
```

Select AMR genes that have a substantial presence in the dataset (n >= 8).
```{r}
pres_abs_top <- pres_abs_general %>% filter(rowSums(.) >= 8)
```

Select the same AMR genes in the matrix with classifications
```{r}
pres_abs_class <- pres_abs_class[rownames(pres_abs_top),]
```

Load phylogeny, we'll use this to order the strains
```{r}
phylogeny <- ape::read.tree("../results/final_ml_tree.newick")
phylogeny$tip.label <- paste0(phylogeny$tip.label,'.fna')
```

Reorder the input pres abs
```{r}
label_order <- phylogeny$tip.label
pres_abs_class <- pres_abs_class[,label_order]
```

Add metadata as annotation
```{r}
metadata <- read.csv("../../rgnosis_metadata/results/Ecoli_metadata_selected.csv") %>% rename(strain=id)
metadata_anno <- metadata %>% mutate(strain=paste0(strain,".fna"))
metadata_anno %<>% slice(match(label_order,strain))
```

Add custom colours for annotation of STs and phylogroups
```{r}
ST_colours <- setNames(c("#f8766d","#ffc107","#00ba38","#00c19f","#00b9e3","#619cff","#db72fb","#979797","#ffffff"),c("10","131","38","410","648","69","88","other","no data"))
phylogroup_colours <- setNames(c("#12939A","#DDB27C","#88572C","#FF991F","#F15C17","#223F9A","#DA70BF","#125C77"),c("A","B1","B2","C","D","E","F","G"))
clade_colours <- setNames(c("#5574f0","#45bd43","#f65e43"),c("A","B","C"))
```

```{r}
colour_class <- c("white","#218f61","#de6909","yellow")
names(colour_class) <- c(0,1,2,3)
```

Cluster rows based on presence / absence of AMR genes (not on whether they were found on the plasmid/chromosome)
```{r}
dist_rows <- proxy::dist(pres_abs_top, by_rows = TRUE, method = "Euclidean")
hclust_rows <- hclust(dist_rows,method = "average")
```

```{r}
anno =HeatmapAnnotation(ST = metadata_anno$ST_group, phylogroup = metadata_anno$phylogroup, col = list(ST=ST_colours,phylogroup=phylogroup_colours))
Heatmap(pres_abs_class,show_column_names = FALSE,top_annotation = anno, cluster_columns = as.dendrogram(phylogeny), cluster_rows = as.dendrogram(hclust_rows), row_dend_reorder = FALSE, col = colour_class, show_row_dend = TRUE, heatmap_legend_param = list(title="predicted contig source",labels=c("","plasmid","chromosome","both")))
```
We can also cluster based on plasmidome
You'll have to import the variable plasmidome_clustering from the panaroo post processing script.
```{r}
#reorder based on plasmidome clustering
label_order <- paste0(gsub("\\.","-",plasmidome_clustering$labels),".fna")
pres_abs_class <- pres_abs_class[,label_order]
ST_reordered <- ST %>% slice(match(label_order,strain))
```

```{r}
ST_anno =HeatmapAnnotation(ST = ST_reordered$ST,col=list(ST=colour_vector))
Heatmap(pres_abs_class,show_column_names = FALSE,top_annotation = ST_anno, cluster_columns = as.dendrogram(plasmidome_clustering), cluster_rows = as.dendrogram(hclust_rows), row_dend_reorder = FALSE, col = colour_class, show_row_dend = TRUE)
```
