---
title: "AMR analysis"
author: "Lisa"
date: "12/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Import/install packages
```{r, warning=FALSE, message=FALSE}
packages <- c("magrittr","dplyr","ggplot2","stringr","cooccur","visNetwork","igraph")
for (package in packages){
  if (!require(package, character.only = TRUE)) install.packages(package)
  require(package, character.only = TRUE)
}
```

Import resfinder file
```{r}
resfinder <- read.delim("../results/ResFinder.tsv")
```

Extract ST131 samples
```{r}
ST131_strains <- read.delim("../results/sample_summaries_ST131.csv",sep = ';') 
resfinder %<>% filter(sub(".fna","",X.FILE) %in% ST131_strains$Sample)
```

Filter based on coverage threshold
```{r}
coverage_threshold <- 95
resfinder %<>% filter(X.COVERAGE >= coverage_threshold)
```

Add information on origin (plasmid/chromosome) of the contigs
```{r}
contig_classifications <- read.csv("../results/EC_combined_output.csv")
resfinder <- left_join(resfinder,contig_classifications %>% select(contig_name,classification),by = c("SEQUENCE"="contig_name"))
```

Count AMR genes per sample
```{r}
count_AMR <- resfinder %>% group_by(X.FILE) %>% summarise(count=n())
```

```{r}
ggplot(count_AMR,aes(x=count))+
  geom_histogram(binwidth=1,stat = 'count')+
  xlim(0,13)+
  scale_x_continuous(breaks = seq(0,13,1))+
  labs(x="Nr. of AMR genes",y="Nr. of samples")
```

Count samples per AMR gene
```{r}
count_samples <- resfinder %>% group_by(GENE,classification) %>% summarise(count=n())
count_samples %<>% group_by(GENE) %>% mutate(total_count=sum(count))
```

```{r}
ggplot(count_samples, aes(x=count,y=reorder(GENE,+total_count), fill=classification))+
  geom_bar(position= 'stack', stat = 'identity')
```

### Co-occurrence analysis
Are certain AMR genes more frequently found together?

Transform the resfinder data to get a presence absence matrix for each gene / sample
```{r}
cooccur_data <- resfinder %>% select(X.FILE,GENE)
pres_abs <- as.data.frame.matrix(table(cooccur_data[2:1]))
```

Use cooccur package to perform co-occurrence analysis
```{r}
cooccurrence <- cooccur(pres_abs, spp_names = TRUE)
cooccur_results <- cooccurrence$results
```
Filter to keep only significant values
```{r}
cooccur_sign <- cooccur_results %>% filter(p_lt <= 0.05 | p_gt <= 0.05)
```

### Visualisation with visNetwork

Build nodes
```{r}
#Find the genes that have significant results (so that are either in s1 or s2 of cooccur_sign)
sign_genes <- (unique(c(cooccur_sign$sp1_name,cooccur_sign$sp2_name)))
#Build noded file with significant genes and their id's (row numbers)
nodes <- data.frame(label=rownames(pres_abs),id=1:nrow(pres_abs)) %>% filter(label %in% sign_genes)
```

Build edges
```{r}
edges <- data.frame(from=cooccur_sign$sp1,to=cooccur_sign$sp2,color=ifelse(cooccur_sign$p_lt <= 0.05,'#D5232F','#4CB942'))
```

Plot the network
```{r}
visNetwork(nodes, edges) %>%
    visIgraphLayout(layout = "layout_with_kk")
```

