---
title: "Panaroo post processing"
author: "Lisa"
date: "30/09/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Import/install packages
```{r, warning=FALSE, message=FALSE}
packages <- c("proxy","magrittr","dplyr","stats","ggplot2","stringr","Rtsne","ape","circlize","RColorBrewer","gplots","pheatmap","phylogram")
for (package in packages){
  if (!require(package, character.only = TRUE)) install.packages(package)
  require(package, character.only = TRUE)
}
```

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
packages <- c("ComplexHeatmap","ggtree")
for (package in packages){
  if (!require(package, character.only = TRUE)) BiocManager::install(package)
  require(package, character.only = TRUE)
}
```

```{r, include=FALSE}
theme_set(theme_bw())
```

Define dataset to analyse: Accessory genome or plasmidome
```{r}
dataset <- "A"
if (dataset == "A") {
  input = "../results/panaroo_output_all/gene_presence_absence.Rtab"
} else if (dataset =="P") {
  input = "../results/plasmidome_all/panaroo_output/gene_presence_absence.Rtab"
}
```


Import gene presence absence matrix.
```{r}
presence_absence <- read.table(input, header = TRUE, row.names = 1)
presence_absence %<>% t(.)   #transpose the data
presence_absence %<>% as.data.frame()
#remove non.Ecoli sample
presence_absence <- presence_absence[row.names(presence_absence) != "ECO.JSC.RGN.103823",]
```

For comparing the strains we are only interested in the accessory genome.
Therefore we delete all core genes, e.g. genes with a 100% presence in the strains (all rows are 1).
```{r}
accessory_presence_absence <- presence_absence %>% select_if(~min(.) == 0)
```

Calculate Jaccard distances
```{r}
jaccard_distances <- proxy::dist(accessory_presence_absence, by_rows = TRUE, method = "Jaccard")
jaccard_matrix <- as.matrix(jaccard_distances)
```

Classical multidimensional scaling
```{r}
mds <- as.data.frame(cmdscale(jaccard_matrix))
colnames(mds) <- c("dim_1","dim_2")
```

Add number of genes as variable to mds data.
```{r}
total_genes <- rowSums(accessory_presence_absence)
mds %<>% mutate(total_genes=total_genes)
```

```{r}
ggplot(mds,aes(x=dim_1,y=dim_2))+
  geom_point(aes(colour=total_genes))+
  scale_colour_gradient(low ="#ffff00", high = "#cc0099")
```

Add metadata
```{r}
metadata <- read.csv("../../rgnosis_metadata/results/Ecoli_metadata_selected.csv") %>% rename(strain=id)
mds %<>% tibble::rownames_to_column("strain") %>% mutate(strain=str_replace_all(strain,fixed('.'),'-'))
mds_metadata <- full_join(mds,metadata)
```
Correlation with genome size?
```{r}
ST131_strains <- read.delim("../results/sample_summaries_ST131.csv",sep = ';') %>% rename(strain=Sample)
mds_metadata <- full_join(mds_metadata,ST131_strains %>% select(strain,Total.length))
```
```{r}
ggplot(mds_metadata,aes(x=Total.length,y=total_genes))+
  geom_point()+
  geom_smooth(method = 'lm', se=FALSE, colour='darkred', size =0.5)
```

```{r}
ggplot(mds_metadata,aes(x=dim_1,y=dim_2))+
  geom_point(aes(colour=ST))
```
```{r}
ggplot(mds_metadata,aes(x=dim_1,y=dim_2))+
  geom_point(aes(colour=treatment))
```

```{r}
ggplot(mds_metadata,aes(x=dim_1,y=dim_2))+
  geom_point(aes(colour=tracti))
```

### Visualisation

Get the MLST
```{r}
MLST <- read.delim("../results/bactofidia_output_all/stats/MLST.tsv",header = FALSE)
colnames(MLST) <- c("strain","species","ST","adk","fumC","gyrB","icd","mdh","purA","recA")
#It also contains some non-Ecoli strains that we are not interested in, let's filter them out
all_ecoli <- read.delim("../results/all_ecoli_rgnosis.txt",header = FALSE)
MLST %<>% filter(sub(".fna","",strain) %in% all_strains$V1)
```

```{r}
#annotation
ST <- MLST %>% select(strain,ST) %>% mutate(strain=gsub("-",".",sub(".fna","",strain)))
ST$ST %<>% as.factor()
colourconverter <- data.frame(
  ST= rev(tail(names(sort(table(ST$ST))),10)),
  colour = c("#ffd618","#3587bb","#46d08d","#ff7252","#c471be","#7d81cd","#f56490","#43a5bb","#00bdb4","#a0da53"))
ST <- merge(ST,colourconverter,by="ST",all.x = TRUE)
ST %<>% mutate(colour=ifelse(is.na(colour),"#979797",colour),colour=ifelse(ST=='-',NA,colour))
```

```{r}
hclust_out <- hclust(jaccard_distances,method = "ward.D2")
label_order <- hclust_out$labels
#reorder ST annotation file
ST_reordered <- ST %>% slice(match(label_order,strain))
```

We can also add the phylogeny instead of the hierarchical clustering
```{r}
phylogeny <- ape::read.tree("../results/final_ml_tree.newick")
dendro <- as.dendrogram(phylogeny)
plot(dendro)
```

```{r}
heatmap(jaccard_matrix,symm = TRUE,Rowv = dendro,labCol = NA,labRow = NA)#,RowSideColors = ST_reordered$colour)
```

If you just do Jaccard, you compare the accessory genome of one strain to the accessory genome of another strain.
If you then calculate Euclidian distance on the Jaccard matrix, you compare all Jaccard distances of one 
strain to all Jaccard distances of another strain. 

