---
title: "Panaroo post processing"
author: "Lisa"
date: "30/09/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Import/install packages
```{r, warning=FALSE, message=FALSE}
packages <- c("proxy","magrittr","dplyr","stats","ggplot2","stringr","Rtsne","ape","circlize","RColorBrewer","gplots","ComplexHeatmap","phylogram","uwot","ggpubr")
for (package in packages){
  if (!require(package, character.only = TRUE)) install.packages(package)
  require(package, character.only = TRUE)
}
```

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
packages <- c("ComplexHeatmap","ggtree")
for (package in packages){
  if (!require(package, character.only = TRUE)) BiocManager::install(package)
  require(package, character.only = TRUE)
}
```

```{r, include=FALSE}
theme_set(theme_bw())
```

Define dataset to analyse: Accessory genome or plasmidome
```{r}
dataset <- "P"
if (dataset == "A") {
  input = "../results/panaroo_output_all/gene_presence_absence.Rtab"
} else if (dataset =="P") {
  input = "../results/plasmidome_all/panaroo_output/gene_presence_absence.Rtab"
}
```

Import gene presence absence matrix.
```{r}
presence_absence <- read.table(input, header = TRUE, row.names = 1)
presence_absence %<>% t(.)   #transpose the data
presence_absence %<>% as.data.frame()
#remove non.Ecoli sample
presence_absence <- presence_absence[row.names(presence_absence) != "ECO.JSC.RGN.103823",]
```

For comparing the strains we are only interested in the accessory genome.
Therefore we delete all core genes, e.g. genes with a 100% presence in the strains (all rows are 1).
Not necessary if doing this analysis for the plasmidome, none of the plasmid genes are core.
```{r}
if (dataset == "A") {
  accessory_presence_absence <- presence_absence %>% select_if(~min(.) == 0)
} 
if (dataset == "P") {
  plasmidome_presence_absence <- presence_absence
}
```

### Analysis of accessory genome and plasmidome size
Plot nr. of genes in accessory genome and plasmidome (first do the previous steps for both dataset A and dataset P)

```{r}
gene_counts <- data.frame(accessory=rowSums(accessory_presence_absence),plasmidome=rowSums(plasmidome_presence_absence))
#divide by ST131 or non-ST131
MLST <- read.delim("../results/bactofidia_output_all/stats/MLST.tsv",header = FALSE)
colnames(MLST) <- c("id","species","ST","adk","fumC","gyrB","icd","mdh","purA","recA")
ST131_ids <- MLST %>% filter(ST=="131") %>% mutate(id=sub(".fna","",gsub("-",".",id)))
gene_counts %<>% rownames_to_column("id") %>% mutate(ST131=ifelse(id %in% ST131_ids$id,"ST131","Other STs"))
#add sample that is ST131 but not recognised by MLST
gene_counts %<>% mutate(ST131=ifelse(id=="ECO.JSC.MOD.104871","ST131",ST131))
```

```{r}
gene_count_plotting <- gene_counts %>% gather("accessory","plasmidome",key ="dataset",value="genes")

# New facet label names for dataset variable
dataset.labs <- c("Accessory genome","Plasmidome")
names(dataset.labs) <- c("accessory","plasmidome")

ggplot(gene_count_plotting,aes(y=genes,x=factor(ST131,levels = c("ST131","Other STs"))))+
  geom_boxplot()+
  facet_wrap(~dataset, labeller = labeller(dataset=dataset.labs))+
  labs(x = NULL)
```

Test whether the difference is significant by t test

Let's see if the variances are equal
```{r}
var.test(genes ~ ST131, accessory_counts)
var.test(genes ~ ST131, plasmidome_counts)
```
They are not equal, we perform a Welch t-test.
```{r}
accessory_counts_ST131 <- gene_count_plotting %>% filter(dataset=="accessory" & ST131=="ST131") %>% select(genes)
accessory_counts_other <- gene_count_plotting %>% filter(dataset=="accessory" & ST131=="Other STs") %>% select(genes)
t.test(accessory_counts_ST131,accessory_counts_other)
```

```{r}
plasmidome_counts_ST131 <- gene_count_plotting %>% filter(dataset=="plasmidome" & ST131=="ST131") %>% select(genes)
plasmidome_counts_other <- gene_count_plotting %>% filter(dataset=="plasmidome" & ST131=="Other STs") %>% select(genes)
t.test(plasmidome_counts_ST131,plasmidome_counts_other)
```
Add significance to the plots

```{r}
gene_count_plotting <- gene_counts %>% gather("accessory","plasmidome",key ="dataset",value="genes")

# New facet label names for dataset variable
dataset.labs <- c("Accessory genome","Plasmidome")
names(dataset.labs) <- c("accessory","plasmidome")

ggplot(gene_count_plotting,aes(y=genes,x=factor(ST131,levels = c("ST131","Other STs")),fill=factor(ST131,levels = c("ST131","Other STs"))))+
  geom_boxplot()+
  facet_wrap(~dataset, labeller = labeller(dataset=dataset.labs))+
  labs(x = NULL)+
  stat_compare_means(method = "t.test",label.y = 1500)+
  theme(legend.position = "none")
```

Plot plasmidome against accessory genome
```{r}
#function for plotting the regression line and R^2
#source= https://groups.google.com/forum/#!topic/ggplot2/1TgH-kG5XM
lm_eqn <- function(df){
    m <- lm(plasmidome ~ accessory, df);
    eq <- substitute(italic(y) == a + b %.% italic(x)*","~~italic(r)^2~"="~r2, 
         list(a = format(unname(coef(m)[1]), digits = 2),
              b = format(unname(coef(m)[2]), digits = 2),
             r2 = format(summary(m)$r.squared, digits = 3)))
    as.character(as.expression(eq));
}
```

```{r}
ggplot(gene_counts,aes(x=accessory,y=plasmidome,colour=factor(ST131,levels = c("ST131","Other STs"))))+
  geom_point()+
  geom_smooth(method = lm, se=F)+
  stat_regline_equation(aes(label = ..eq.label..))+
  stat_regline_equation(label.x = 2000, aes(label = ..rr.label..))+
  labs(colour="Dataset",y="Nr. of genes in plasmidome",x="Nr. of genes in accessory genome")
```

### Calculate and visualise distances

Calculate Jaccard distances
```{r}
if (dataset == "A") {
  pres_abs <- accessory_presence_absence
} else if (dataset == "P") {
  pres_abs <- plasmidome_presence_absence
}
jaccard_distances <- proxy::dist(pres_abs, by_rows = TRUE, method = "Jaccard")
jaccard_matrix <- as.matrix(jaccard_distances)
```

Build metadata
```{r}
#first add the total nr. of genes found
metadata <- data.frame(total_genes=rowSums(pres_abs))
metadata %<>% rownames_to_column("strain")
```

Add patient metadata
```{r}
patient_metadata <- read.csv("../../rgnosis_metadata/results/Ecoli_metadata_selected.csv") %>% rename(strain=id)
patient_metadata %<>% mutate(strain=str_replace_all(strain,fixed('-'),'.'))
metadata %<>% full_join(patient_metadata, by = "strain")
```

There are different techniques for reducing the dimensionality of our data (our Jaccard matrix)
MDS is a relatively simple technique which mainly shows global structure.
t-SNE is very good at showing local structure, less so the global structure, although it can be changed with the perplexity parameter.
The cluster sizes and distances you see in t-SNE don't have meaning.
An alternative to t-SNE is UMAP, which is better at preserving global structure.

Classical multidimensional scaling
```{r}
mds <- as.data.frame(cmdscale(jaccard_matrix))
colnames(mds) <- c("dim_1","dim_2")
```

Add metadata
```{r}
mds %<>% merge(metadata,by.x = 'row.names', by.y = 'strain')
```

```{r}
ggplot(mds,aes(x=dim_1,y=dim_2,colour=ST_group))+
  geom_point()+
  stat_ellipse(data=mds %>% filter(!ST_group %in% c('no data','other')))+
  scale_colour_manual(values=c("#f8766d","#ffc107","#00ba38","#00c19f","#00b9e3","#619cff","#db72fb","#BCBCBC","#000000"))
```

t-SNE

```{r}
#calculate tSNE coordinates
tSNE <- as.data.frame(Rtsne(jaccard_distances,is_distance = TRUE,theta = 0.0,perplexity = 30)$Y)
colnames(tSNE) <- c("dim_1","dim_2")
#add strain names
tSNE %<>% mutate(strain=rownames(jaccard_matrix))
```

Add metadata
```{r}
tSNE %<>% merge(metadata)
```

```{r}
ggplot(tSNE,aes(x=dim_1,y=dim_2,colour=ST_group))+
  geom_point()+
  stat_ellipse(data=tSNE %>% filter(!ST_group %in% c('no data','other')))+
  scale_colour_manual(values=c("#f8766d","#ffc107","#00ba38","#00c19f","#00b9e3","#619cff","#db72fb","#BCBCBC","#000000"))
```

```{r}
ggplot(tSNE,aes(x=dim_1,y=dim_2,colour=total_genes))+
  geom_point()
```

UMAP

```{r}
umap <- umap(jaccard_distances,n_neighbors=50)
colnames(umap) <- c("dim_1","dim_2")
```

Add metadata
```{r}
umap %<>% merge(metadata,by.x = 'row.names', by.y = 'strain')
```

```{r}
ggplot(umap,aes(x=dim_1,y=dim_2,colour=ST_group))+
  geom_point()+
  stat_ellipse(data=umap %>% filter(!ST_group %in% c('no data','other')))+
  scale_colour_manual(values=c("#f8766d","#ffc107","#00ba38","#00c19f","#00b9e3","#619cff","#db72fb","#BCBCBC","#000000"))
```

### Heatmap

Cluster strains based on jaccard distances.
Average linking conserves the structures we see in the dimensionality reduction plots best.
```{r}
hclust_out <- hclust(jaccard_distances,method = "average")
label_order <- hclust_out$labels
#reorder ST annotation file
ST_reordered <- ST %>% slice(match(label_order,strain))
```

Save hclust output 
```{r}
if (dataset == "P") {
  plasmidome_clustering <- hclust_out
}
```

Add metadata as annotation
```{r}
metadata_anno <- metadata %>% mutate(strain=gsub("-",".",sub(".fna","",strain)))
#There is one strain for which metadata is missing, let's add it manually
row_i <- nrow(metadata_anno)+1
metadata_anno[row_i,] <- NA
metadata_anno[row_i,1] <- "ECO.MSA.BAC.105531"
metadata_anno[row_i,41] <- 224
metadata_anno %<>% slice(match(label_order,strain))
```

Add custom colours for annotation of STs and phylogroups
```{r}
ST_colours <- setNames(c("#f8766d","#ffc107","#00ba38","#00c19f","#00b9e3","#619cff","#db72fb","#979797","#ffffff"),c("10","131","38","410","648","69","88","other","no data"))
phylogroup_colours <- setNames(c("#12939A","#DDB27C","#88572C","#FF991F","#F15C17","#223F9A","#DA70BF","#125C77"),c("A","B1","B2","C","D","E","F","G"))
```

We can also add the phylogeny instead of the hierarchical clustering (optional)
```{r}
phylogeny <- ape::read.tree("../results/final_ml_tree.newick")
phylogeny$tip.label <- gsub("-",".",phylogeny$tip.label)
dendro <- as.dendrogram(phylogeny)
```

```{r}
col_fun = colorRamp2(c(0,1), c("#fff7d9", "#d12110"))
anno =HeatmapAnnotation(ST = metadata_anno$ST_group, phylogroup = metadata_anno$phylogroup, col = list(ST=ST_colours,phylogroup=phylogroup_colours))
Heatmap(jaccard_matrix, heatmap_width = unit(20, "cm"), heatmap_height = unit(20, "cm"),show_row_names = FALSE, show_column_names = FALSE,col = col_fun, cluster_rows = as.dendrogram(hclust_out), cluster_columns = as.dendrogram(hclust_out),name='Jaccard distance', top_annotation = anno)
```


If you just do Jaccard, you compare the accessory genome of one strain to the accessory genome of another strain.
If you then calculate Euclidian distance on the Jaccard matrix, you compare all Jaccard distances of one 
strain to all Jaccard distances of another strain. 

