---
title: "st_131_plasmid_quast"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(echo = TRUE)
```

```{r}
require(readr)
require(tidyr)
require(dplyr)
require(magrittr)
require(ggplot2)
require(stringr)

```

```{r, include=FALSE}
theme_set(theme_bw())
```

## Script for processing QUAST output

Get a list of all replicons, with their references_id and the lengths
```{r}
replicon_data <- read.csv("../../../ST131_ncbi_download/results/replicon_data.csv", header = FALSE)
names(replicon_data)<-c('assembly_accession','reference_id','replicon_length','classification')
```

Import and process the results from QUAST
```{r}
#MOB
quast_output_mob<-read.csv('../../mobsuite/results/quast_output/mob_alignments_statistics.csv', sep=',', header=FALSE)
names(quast_output_mob)<-c('assembly_accession','sra_accession','bin_name_temp','reference_id','alignment_length','bin_length')
quast_output_mob %<>% mutate(software = "MOB-suite (unicycler, untrimmed)")
#For MOB-suite we will fuse the name of the bin with the name of the strain, to be able to do statistics properly
quast_output_mob<-unite(quast_output_mob,'bin_name',c('sra_accession','bin_name_temp'),remove=FALSE) 
#remove old bin name
quast_output_mob<-quast_output_mob[,c(2,1,3,5:8)]

#MOB_BAC
quast_output_mob_bac<-read.csv('../../mobsuite_bac/results/quast_output/mob_alignments_statistics.csv', sep=',', header=FALSE)
names(quast_output_mob_bac)<-c('assembly_accession','sra_accession','bin_name_temp','reference_id','alignment_length','bin_length')
quast_output_mob_bac %<>% mutate(software = "MOB-suite (bactofidia)")
#For MOB-suite we will fuse the name of the bin with the name of the strain, to be able to do statistics properly
quast_output_mob_bac<-unite(quast_output_mob_bac,'bin_name',c('sra_accession','bin_name_temp'),remove=FALSE) 
#remove old bin name
quast_output_mob_bac<-quast_output_mob_bac[,c(2,1,3,5:8)]

#MOB_TRIMMED
quast_output_mob_trimmed<-read.csv('../../mobsuite_trimmed/results/quast_output/mob_alignments_statistics.csv', sep=',', header=FALSE)
names(quast_output_mob_trimmed)<-c('assembly_accession','sra_accession','bin_name_temp','reference_id','alignment_length','bin_length')
quast_output_mob_trimmed %<>% mutate(software = "MOB-suite (unicycler, trimmed)")
#For MOB-suite we will fuse the name of the bin with the name of the strain, to be able to do statistics properly
quast_output_mob_trimmed<-unite(quast_output_mob_trimmed,'bin_name',c('sra_accession','bin_name_temp'),remove=FALSE) 
#remove old bin name
quast_output_mob_trimmed<-quast_output_mob_trimmed[,c(2,1,3,5:8)]

#MOB_EC
quast_output_mob_ec<-read.csv('../../mobsuite_EC/results/quast_output/mob_alignments_statistics.csv', sep=',', header=FALSE)
names(quast_output_mob_ec)<-c('assembly_accession','sra_accession','bin_name_temp','reference_id','alignment_length','bin_length')
quast_output_mob_ec %<>% mutate(software = "MOB-suite (bactofidia, plasmid contigs)")
#For MOB-suite we will fuse the name of the bin with the name of the strain, to be able to do statistics properly
quast_output_mob_ec<-unite(quast_output_mob_ec,'bin_name',c('sra_accession','bin_name_temp'),remove=FALSE) 
#remove old bin name
quast_output_mob_ec<-quast_output_mob_ec[,c(2,1,3,5:8)]

#MOB_CLEAN
quast_output_mob_clean<-read.csv('../../mobsuite_clean/results/quast_output_clean/mob_alignments_statistics.csv', sep=',', header=FALSE)
names(quast_output_mob_clean)<-c('assembly_accession','sra_accession','bin_name_temp','reference_id','alignment_length','bin_length')
quast_output_mob_clean %<>% mutate(software = "MOB-suite (bactofidia, cleaned)")
#For MOB-suite we will fuse the name of the bin with the name of the strain, to be able to do statistics properly
quast_output_mob_clean<-unite(quast_output_mob_clean,'bin_name',c('sra_accession','bin_name_temp'),remove=FALSE) 
#remove old bin name
quast_output_mob_clean<-quast_output_mob_clean[,c(2,1,3,5:8)]


#SPADES
quast_output_spades<-read.csv('../../plasmidspades/results/quast_output/spades_alignments_statistics.csv', sep = ',', header = FALSE)
names(quast_output_spades)<-c('assembly_accession','sra_accession','bin_name','reference_id','alignment_length','bin_length')
quast_output_spades %<>% mutate(software = "PlasmidSPAdes")
```

Combine all quast output and cross with replicon data
```{r}
quast_output_all<-rbind(quast_output_mob,quast_output_mob_bac,quast_output_mob_trimmed,quast_output_spades,quast_output_mob_ec,quast_output_mob_clean)
quast_output_plasmids_chromosomes<-full_join(replicon_data,quast_output_all,by=c('assembly_accession','reference_id'))
```

```{r}

```

Filter out duplicated strains.
```{r}
quast_output_filtered<-quast_output_plasmids_chromosomes %>% filter(!sra_accession %in% c("SRR4025862","SRR5168233","SRR4025855","SRR5168210","SRR4025854","SRR5167858"))
```


For all bins that align to a plasmid, we calculate the recall (percentage of plasmid bps included in the bin) and precision (percentage of bps in the bin that are plasmid derived).

```{r}
#Filter replicons classified as plasmids
quast_output_plasmids<-filter(quast_output_filtered,classification=='plasmid')
#For cases in which the bin_name was not present, this means that the tool did not detected this plasmid. Assign that category.
quast_output_plasmids$bin_name<-ifelse(is.na(quast_output_plasmids$bin_name),"not_detected",as.character(quast_output_plasmids$bin_name))
#For this cases also assign an alignment length =0
quast_output_plasmids$alignment_length[is.na(quast_output_plasmids$alignment_length)]<-0
#assign not applicable to all the rest of NA
quast_output_plasmids[is.na(quast_output_plasmids)]<-'not_applicable'
```

### Calculating Recall
```{r}
quast_output_plasmids$recall_bp<-as.numeric(quast_output_plasmids$alignment_length)/as.numeric(quast_output_plasmids$replicon_length)
```

Now assign a 'not detected' category for recall=0 (to exclude this values from calculations
```{r}
quast_output_plasmids$recall_bp<-ifelse(quast_output_plasmids$recall_bp==0,'not_detected',as.numeric(quast_output_plasmids$recall_bp))
```

### Calculating Precision
For calculating precision correctly, we will have to substract the exceeding multiple alignments lengths. (otherwise, total alignment could be larger than the bin-length, therefore precision could be larger than 1, and that is not possible) 

To achieve this, we will first obtain how many times each contig aligns to the replicon (n), then we would substract one from that number (n-1) and finally we will substract (n-1)*contig_length from the alignment length for those specific cases.

n, and contig_length can be obtained from the files 'mob_ambiguous_references', which was an output from the 'gather_quast_results.py script'

Import Ambiguous alignment information and wrangle the data
```{r}
#mob
ambiguous_alignments_mob<-read.csv('../../mobsuite/results/quast_output/mob_ambiguous_references.csv', sep=',', header=FALSE)
names(ambiguous_alignments_mob)<-c('sra_accession','bin_name_temp','reference_id','contig_name','contig_length','start','end')
ambiguous_alignments_mob<-unite(ambiguous_alignments_mob,'bin_name',c('sra_accession','bin_name_temp'),remove=FALSE) 
#remove old bin name
ambiguous_alignments_mob<-ambiguous_alignments_mob[,c(1,2,4:8)]
ambiguous_alignments_mob$software<-'MOB-suite (unicycler, untrimmed)'

#mob_bact
ambiguous_alignments_mob_bac<-read.csv('../../mobsuite_bac/results/quast_output/mob_ambiguous_references.csv', sep=',', header=FALSE)
names(ambiguous_alignments_mob_bac)<-c('sra_accession','bin_name_temp','reference_id','contig_name','contig_length','start','end')
ambiguous_alignments_mob_bac<-unite(ambiguous_alignments_mob_bac,'bin_name',c('sra_accession','bin_name_temp'),remove=FALSE) 
#remove old bin name
ambiguous_alignments_mob_bac<-ambiguous_alignments_mob_bac[,c(1,2,4:8)]
ambiguous_alignments_mob_bac$software<-'MOB-suite (bactofidia)'

#mob_trimmed
ambiguous_alignments_mob_trimmed<-read.csv('../../mobsuite_trimmed/results/quast_output/mob_ambiguous_references.csv', sep=',', header=FALSE)
names(ambiguous_alignments_mob_trimmed)<-c('sra_accession','bin_name_temp','reference_id','contig_name','contig_length','start','end')
ambiguous_alignments_mob_trimmed<-unite(ambiguous_alignments_mob_trimmed,'bin_name',c('sra_accession','bin_name_temp'),remove=FALSE) 
#remove old bin name
ambiguous_alignments_mob_trimmed<-ambiguous_alignments_mob_trimmed[,c(1,2,4:8)]
ambiguous_alignments_mob_trimmed$software<-'MOB-suite (unicycler, trimmed)'

#mob_ec
ambiguous_alignments_mob_ec<-read.csv('../../mobsuite_EC/results/quast_output/mob_ambiguous_references.csv', sep=',', header=FALSE)
names(ambiguous_alignments_mob_ec)<-c('sra_accession','bin_name_temp','reference_id','contig_name','contig_length','start','end')
ambiguous_alignments_mob_ec<-unite(ambiguous_alignments_mob_ec,'bin_name',c('sra_accession','bin_name_temp'),remove=FALSE) 
#remove old bin name
ambiguous_alignments_mob_ec<-ambiguous_alignments_mob_ec[,c(1,2,4:8)]
ambiguous_alignments_mob_ec$software<-'MOB-suite (bactofidia, plasmid contigs)'

#mob_clean
ambiguous_alignments_mob_clean<-read.csv('../../mobsuite_clean/results/quast_output_clean/mob_ambiguous_references.csv', sep=',', header=FALSE)
names(ambiguous_alignments_mob_clean)<-c('sra_accession','bin_name_temp','reference_id','contig_name','contig_length','start','end')
ambiguous_alignments_mob_clean<-unite(ambiguous_alignments_mob_clean,'bin_name',c('sra_accession','bin_name_temp'),remove=FALSE) 
#remove old bin name
ambiguous_alignments_mob_clean<-ambiguous_alignments_mob_clean[,c(1,2,4:8)]
ambiguous_alignments_mob_clean$software<-'MOB-suite (bactofidia, cleaned)'

#spades
ambiguous_alignments_spades<-read.csv('../../plasmidspades/results/quast_output/spades_ambiguous_references.csv', sep=',', header=FALSE)
names(ambiguous_alignments_spades)<-c('sra_accession','bin_name','reference_id','contig_name','contig_length','start','end')
ambiguous_alignments_spades$software<-'PlasmidSPAdes'
```


```{r, warning=FALSE, fig.width=14, fig.height=18}
#merge files of different softwares
ambiguous_alignments_total<-rbind(ambiguous_alignments_mob,ambiguous_alignments_mob_bac,ambiguous_alignments_mob_trimmed,ambiguous_alignments_spades,ambiguous_alignments_mob_ec,ambiguous_alignments_mob_clean)

#remove duplications of the same ambiguous contig being included twice
ambiguous_alignments_total<-unique(ambiguous_alignments_total)

#count how many times each contig ambiguously aligns to the same replicon (n)
count_ambiguous_alignments_to_same_replicon<- ambiguous_alignments_total %>% group_by(sra_accession,bin_name,reference_id,contig_name,software) %>% summarise(total_ambiguous_alignments=n())

#join this information with the ambiguous_alignments
all_ambiguous_alignments<-inner_join(ambiguous_alignments_total,count_ambiguous_alignments_to_same_replicon,by=c('sra_accession','bin_name','reference_id','contig_name','software'))

#filter out the start and end column
all_ambiguous_alignments<-all_ambiguous_alignments[,-c(6,7)]

#filter the multiple_alignments 
multiple_ambiguous_alignments<-unique(filter(all_ambiguous_alignments,total_ambiguous_alignments>=2))
#the problem here is that we have several entries that have a slightly different contig lenght. We will retain the largest
multiple_ambiguous_alignments<-multiple_ambiguous_alignments %>% group_by(sra_accession,bin_name,reference_id,contig_name,software) %>% mutate(max_length=max(contig_length)) %>% ungroup() %>% filter(contig_length==max_length)

#get the total substraciton contigs (n-1)
multiple_ambiguous_alignments$substraction_times<-multiple_ambiguous_alignments$total_ambiguous_alignments-1

#now get the length that we will have to substract
multiple_ambiguous_alignments$ambiguous_substraction_length<-multiple_ambiguous_alignments$contig_length*multiple_ambiguous_alignments$substraction_times

#now we will sum all ambiguous substraction lengths that belong to the same bin_name/refrence_id (because there might be multiple contigs within the same bin that align to the same replicon)
multiple_ambiguous_alignments_final_substracion_data<-multiple_ambiguous_alignments %>% group_by(sra_accession,bin_name,reference_id,software) %>% summarise(total_substraction_length=sum(ambiguous_substraction_length))
 

#2.---- Cross information with the 'quast_output_plasmids' from previous section (this will allow us to calculate precision) ------
quast_output_plasmids<-left_join(quast_output_plasmids,multiple_ambiguous_alignments_final_substracion_data,by=c('sra_accession','bin_name','reference_id','software'))
#replace NA values with 0 in the total_substraction_length (for cases in which there are no multiple alignments)
quast_output_plasmids$total_substraction_length[is.na(quast_output_plasmids$total_substraction_length)]<-0

#2.1 Substract 'full alignment length' with 'multiple alignment length' to get the 'alignment length' that will be used for precision calculation
quast_output_plasmids$alignment_length_precision<-(quast_output_plasmids$alignment_length-quast_output_plasmids$total_substraction_length)


#3. ------- Calculate precision
quast_output_plasmids$precision_bp<-quast_output_plasmids$alignment_length_precision/as.numeric(quast_output_plasmids$bin_length)

#format the data and round the results
quast_output_plasmids$precision_bp<-as.numeric(quast_output_plasmids$precision_bp)
quast_output_plasmids$recall_bp<-as.numeric(quast_output_plasmids$recall_bp)
quast_output_plasmids$recall_bp<-round(quast_output_plasmids$recall_bp,3)
quast_output_plasmids$precision_bp<-round(quast_output_plasmids$precision_bp,3)


#4 --- Calculate f1-score
quast_output_plasmids$f1_score<-round(2*((quast_output_plasmids$precision_bp*quast_output_plasmids$recall_bp)/(quast_output_plasmids$precision_bp+quast_output_plasmids$recall_bp)),3)

#convert NA to 0 for scores
quast_output_plasmids[is.na(quast_output_plasmids)]<-0

```
Analyse how many bins are aligned to either plasmid, chromosome, plasmid and chromosome or not aligned at all.
```{r}
bin_analysis <- quast_output_filtered %>% filter(!is.na(bin_name)) %>% group_by(software) %>% mutate(multiple_replicons=duplicated(bin_name) | duplicated(bin_name, fromLast = TRUE))
bin_analysis %<>% mutate(bin_type=ifelse(multiple_replicons==FALSE & classification=="chromosome","chromosome",NA))
bin_analysis %<>% mutate(bin_type=ifelse(multiple_replicons==FALSE & classification=="plasmid","one_plasmid",bin_type))
bin_analysis %<>% group_by(software,bin_name) %>% mutate(bin_type=ifelse(multiple_replicons==TRUE & n_distinct(classification)==1,"multiple_plasmids",bin_type))
bin_analysis %<>% group_by(software,bin_name) %>% mutate(bin_type=ifelse(multiple_replicons==TRUE & n_distinct(classification)!=1,"plasmid_and_chromosome",bin_type))
bin_analysis %<>% mutate(bin_type=ifelse(reference_id=="no_correct_alignments","no_correct_alignments",bin_type))

bin_analysis %<>% ungroup()
count_bin_types <- bin_analysis %>% select(bin_name,software,bin_type) %>% unique()
```

```{r}
level_order_types <- c("one_plasmid","multiple_plasmids","plasmid_and_chromosome","chromosome","no_correct_alignments")
ggplot(count_bin_types,aes(y=factor(bin_type,level = rev(level_order_types)),fill=factor(bin_type,level = rev(level_order_types))))+
  geom_bar()+
  facet_wrap(~software)+
  labs(x = "Number of bins", y = "Bin content")+
  scale_fill_manual(values=rev(c("#75C24B","#47954D","#fc8e13","#eb5a48","#5c5d5f")))+
  scale_y_discrete(labels=rev(c("One plasmid","Multiple plasmids","Plasmid and chromosome","Chromosome","No correct alignments")))+
  guides(fill= FALSE)
```

What's up with the increase in plasmid/chromosome mixed bins when using the ensemble classifier?
```{r}
multiple_plasmids <- count_bin_types %>% filter(software=="MOB-suite (bactofidia, cleaned)" | software=="MOB-suite (bactofidia, plasmid contigs)") %>% filter(bin_type=="multiple_plasmids" )
write.csv(multiple_plasmids,"../results/multiple_plasmids.csv",row.names = FALSE)
```

Composition of mixed bins
```{r}
mixed_bin_analysis <- bin_analysis %>% filter(classification=="chromosome") #& multiple_replicons==TRUE)
mixed_bin_analysis %<>% left_join(multiple_ambiguous_alignments_final_substracion_data,by=c('sra_accession','bin_name','reference_id','software'))
#replace NA values with 0 in the total_substraction_length (for cases in which there are no multiple alignments)
mixed_bin_analysis$total_substraction_length[is.na(mixed_bin_analysis$total_substraction_length)]<-0

#update alignment length
mixed_bin_analysis %<>% mutate(actual_alignment_length=alignment_length-total_substraction_length)
mixed_bin_analysis %<>% mutate(chromosomal_content=actual_alignment_length/bin_length*100)
```

```{r}
ggplot(mixed_bin_analysis,aes(x=chromosomal_content,fill=bin_type))+
  geom_histogram(binwidth = 5,breaks= seq(0,100,by=5))+
  scale_x_continuous(breaks= seq(0,100,by=20))+
  facet_wrap(~software)
```
Find out whether a reference plasmid was predicted by the four methods, and whether it was split over bins.

```{r}
bin_number_plotting<-quast_output_plasmids %>% group_by(reference_id) %>% count(software)
not_detected<-quast_output_plasmids %>% filter(software=='not_applicable') %>% select(reference_id) %>% uncount(6)
not_detected %<>% mutate(software = rep(c("MOB-suite (bactofidia)","MOB-suite (unicycler, trimmed)","MOB-suite (unicycler, untrimmed)","PlasmidSPAdes","MOB-suite (bactofidia, cleaned)","MOB-suite (bactofidia, plasmid contigs)"),times=nrow(not_detected)/6)) %>% mutate(n = 0)
bin_number_plotting<-rbind(bin_number_plotting,not_detected)
bin_number_plotting %<>% spread(software,n) %>% select(!not_applicable)
bin_number_plotting[is.na(bin_number_plotting)]<-0
bin_number_plotting %<>% gather('software','n',2:7)
bin_number_plotting %<>% mutate(correct=n==1)
```

```{r}
ggplot(bin_number_plotting,aes(x=reference_id,y=n,group=1,colour=correct))+
  geom_point()+
  facet_wrap(~software)+
  scale_y_continuous(breaks= seq(0,10,by=2))+
  guides(colour=FALSE)+
  scale_x_discrete(labels=NULL,breaks=NULL)
```

Some plotting
```{r}
length_data <- replicon_data %>% filter(classification=="plasmid")
ggplot(length_data,aes(x=log(replicon_length,10)))+geom_histogram(binwidth = 0.05)
```

```{r}
plotting_data <- filter(quast_output_plasmids,software!="not_applicable")
plotting_data$software = str_wrap(plotting_data$software, width=20)
ggplot(plotting_data,aes(x=software,y=f1_score)) +geom_boxplot() +geom_jitter(shape=16, position=position_jitter(0.2))
```
```{r}
plotting_data <- filter(quast_output_plasmids,software!="not_applicable")
plotting_data$software = str_wrap(plotting_data$software, width=20)
ggplot(plotting_data,aes(x=software,y=precision_bp)) +geom_boxplot() +geom_jitter(shape=16, position=position_jitter(0.2))
```

```{r}
plotting_data <- filter(quast_output_plasmids,software!="not_applicable")
plotting_data$software = str_wrap(plotting_data$software, width=20)
ggplot(plotting_data,aes(x=software,y=recall_bp)) +geom_boxplot() +geom_jitter(shape=16, position=position_jitter(0.2))
```

```{r}
ggplot(plotting_data,aes(x=recall_bp,y=precision_bp)) +
  geom_count() +
  scale_size_continuous(range = c(1.5,5)) +
  facet_wrap(~software)
```