---
title: "Ecoli_plasmid_reconstruction_Benchmark"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(echo = TRUE)
```

Load packages
```{r}
packages<-c("readr","tidyr","dplyr","magrittr","ggplot2","stringr","packcircles","patchwork")
for (package in packages){
  if (!require(package, character.only = TRUE)) install.packages(package)
  require(package, character.only = TRUE)
}
```

Use a nice theme for ggplot
```{r, include=FALSE}
theme_set(theme_bw())
```

## Script for processing QUAST output

Get a list of all replicons, with their references_id and the lengths
```{r}
replicon_data <- read.csv("../../../Ecoli_ncbi_download/results/replicon_data.csv", header = FALSE)
names(replicon_data)<-c('strain','reference_id','replicon_length','classification')
```

Import and process the results from QUAST

I defined a function for importing the MOB-suite results because I needed to execute the same code many times:
```{r}
import_quast_mob <- function(mode) {
  path<-paste0('../../plasmid_reconstruction/results/Ecoli/quast_statistics/',mode,'/',mode,'_alignments_statistics.csv')
  df<-read.csv(path, sep=',', header=FALSE)
  names(df)<-c('strain','bin_name_temp','reference_id','alignment_length','bin_length')
  df %<>% mutate(software = mode)
  #For MOB-suite we will fuse the name of the bin with the name of the strain, to be able to do statistics   properly
  df<-unite(df,'bin_name',c('strain','bin_name_temp'),remove=FALSE) 
  #remove old bin name
  df<-df[,c(2,1,4:7)]
  return(df)
}
```

Run the function for all mobsuite modes:
```{r}
modes <- c("mob_uni","mob_unitrim","mob_bac","mob_uni_cleaned","mob_uni_filtered","mob_bac_cleaned","mob_bac_filtered")

for (mode in modes){
  #import quast results
  quast_out <- import_quast_mob(mode)
  #rename the dataframe
  name <- paste0("quast_output_",mode)
  assign(name,quast_out)
}

```

Import PlasmidSPAdes results:
```{r}
#SPADES
quast_output_spades<-read.csv('../../plasmid_reconstruction/results/Ecoli/quast_statistics/spades/spades_alignments_statistics.csv', sep = ',', header = FALSE)
names(quast_output_spades)<-c('strain','bin_name','reference_id','alignment_length','bin_length')
quast_output_spades %<>% mutate(software = "spades")
```

Combine all quast output and cross with replicon data
```{r}
quast_output_all<-rbind(quast_output_mob_bac,quast_output_mob_bac_cleaned,quast_output_mob_bac_filtered,quast_output_mob_uni,quast_output_mob_uni_cleaned,quast_output_mob_uni_filtered)
quast_output_plasmids_chromosomes<-full_join(replicon_data,quast_output_all,by=c('strain','reference_id'))
```

For all bins that align to a plasmid, we calculate the recall (percentage of plasmid bps included in the bin) and precision (percentage of bps in the bin that are plasmid derived).


### Calculating Precision
For calculating precision correctly, we will have to substract the exceeding multiple alignments lengths. (otherwise, total alignment could be larger than the bin-length, therefore precision could be larger than 1, and that is not possible) 

To achieve this, we will first obtain how many times each contig aligns to the replicon (n), then we would substract one from that number (n-1) and finally we will substract (n-1)*contig_length from the alignment length for those specific cases.

n, and contig_length can be obtained from the files 'mob_ambiguous_references', which was an output from the 'gather_quast_results.py script'

Import Ambiguous alignment information and wrangle the data
```{r}
import_ambiguous_alignments_mob <- function(mode){
  path <- paste0('../../plasmid_reconstruction/results/Ecoli/quast_statistics/',mode,'/',mode,'_ambiguous_references.csv')
  df<-read.csv(path, sep=',', header=FALSE)
  names(df)<-c('strain','bin_name_temp','reference_id','contig_name','contig_length','start','end')
  df<-unite(df,'bin_name',c('strain','bin_name_temp'),remove=FALSE) 
  #remove old bin name
  df<-df[,c(1,2,4:8)]
  df$software<-mode
  return(df)
}
```

```{r}
modes <- c("mob_uni","mob_unitrim","mob_bac","mob_uni_cleaned","mob_uni_filtered","mob_bac_cleaned","mob_bac_filtered")

for (mode in modes){
  #import quast results
  amb_out <- import_ambiguous_alignments_mob(mode)
  #rename the dataframe
  name <- paste0("ambiguous_alignments_",mode)
  assign(name,amb_out)
}
```

```{r}
#spades
ambiguous_alignments_spades<-read.csv('../../plasmid_reconstruction/results/Ecoli/quast_statistics/spades/spades_ambiguous_references.csv', sep=',', header=FALSE)
names(ambiguous_alignments_spades)<-c('strain','bin_name','reference_id','contig_name','contig_length','start','end')
ambiguous_alignments_spades$software<-'spades'
```


```{r, warning=FALSE, fig.width=14, fig.height=18}
#merge files of different softwares
ambiguous_alignments_total<-rbind(ambiguous_alignments_mob_bac,ambiguous_alignments_mob_bac_cleaned,ambiguous_alignments_mob_bac_filtered,ambiguous_alignments_mob_uni,ambiguous_alignments_mob_uni_cleaned,ambiguous_alignments_mob_uni_filtered)

#remove duplications of the same ambiguous contig being included twice
ambiguous_alignments_total<-unique(ambiguous_alignments_total)

#count how many times each contig ambiguously aligns to the same replicon (n)
count_ambiguous_alignments_to_same_replicon<- ambiguous_alignments_total %>% group_by(strain,bin_name,reference_id,contig_name,software) %>% summarise(total_ambiguous_alignments=n(),ambiguous_alignment=sum(contig_length))

#join this information with the ambiguous_alignments
all_ambiguous_alignments<-inner_join(ambiguous_alignments_total,count_ambiguous_alignments_to_same_replicon,by=c('strain','bin_name','reference_id','contig_name','software'))

#filter out the start and end column
all_ambiguous_alignments<-all_ambiguous_alignments[,-c(6,7)]
```


```{r, warning=FALSE, fig.width=14, fig.height=18}
#the problem here is that we have several entries that have a slightly different contig lenght. We will calculate the median
all_ambiguous_alignments %<>% group_by(strain,bin_name,reference_id,contig_name,software) %>% mutate(contig_length=round(median(contig_length))) %>% ungroup()
#keep only unique rows
all_ambiguous_alignments %<>% unique()

#get the total substraciton contigs (n-1)
all_ambiguous_alignments$substraction_times<-all_ambiguous_alignments$total_ambiguous_alignments-1

#now get the length that we will have to substract
all_ambiguous_alignments$ambiguous_substraction_length<-all_ambiguous_alignments$contig_length*all_ambiguous_alignments$substraction_times

#now we will sum all ambiguous substraction lengths that belong to the same bin_name/refrence_id (because there might be multiple contigs within the same bin that align to the same replicon)
all_ambiguous_alignments_final_substracion_data<-all_ambiguous_alignments %<>% group_by(strain,bin_name,reference_id,software) %>% summarise(total_substraction_length=sum(ambiguous_substraction_length),total_ambiguous_length=sum(ambiguous_alignment))
```

```{r, warning=FALSE, fig.width=14, fig.height=18}
#2.---- Cross information with the 'quast_output_plasmids_chromosomes' from previous section (this will allow us to calculate precision) ------
quast_output_plasmids_chromosomes<-left_join(quast_output_plasmids_chromosomes,all_ambiguous_alignments_final_substracion_data,by=c('strain','bin_name','reference_id','software'))
#replace NA values with 0 in the total_substraction_length (for cases in which there are no multiple alignments)
quast_output_plasmids_chromosomes$total_substraction_length[is.na(quast_output_plasmids_chromosomes$total_substraction_length)]<-0
quast_output_plasmids_chromosomes$total_ambiguous_length[is.na(quast_output_plasmids_chromosomes$total_ambiguous_length)]<-0

#Substract 'full alignment length' with 'multiple alignment length' to get the 'alignment length' that will be used for precision calculation
quast_output_plasmids_chromosomes$alignment_length_precision<-(quast_output_plasmids_chromosomes$alignment_length-quast_output_plasmids_chromosomes$total_substraction_length)
```

If the ONLY alignment of a bin to a replicon is ambiguous, this alignment does not count!
A bin which includes a repeated element might map to a replicon by only this repeated element. This makes it seems like it merges replicons together, while this isn't the case!

```{r, warning=FALSE, fig.width=14, fig.height=18}
#Calculate the non-ambiguous length.
#I noticed that sometimes the non-ambiguous length is negative, which shouldn't happen. The ambiguous alignments don't overlap completely with the alignment length in the alignment statistics file, probably a bug in the gather_quast_results script. In one case there was a contig included in the ambiguous file but not in the alignment statistics file.
#Because it was very infrequent, I removed alignments with non-ambiguous length >= 0.
quast_output_plasmids_chromosomes %<>% mutate(non_ambiguous_length=alignment_length-total_ambiguous_length)


#How many fully ambiguous bins do we have? Because removing these could be problematic, t removes the whole bin from the dataset!
explore_fully_ambiguous <- quast_output_plasmids_chromosomes %>% group_by(bin_name,software) %>% summarise(total_align=sum(alignment_length),total_amb=sum(total_ambiguous_length))
explore_fully_ambiguous %<>% mutate(fully_ambiguous=total_align==total_amb) %>% filter(!total_align==0)

quast_output_plasmids_chromosomes %<>% left_join(.,explore_fully_ambiguous %>% select(!c(total_align,total_amb)))

#Calculate number of alignments that would be removed, and how many of them belong to fully ambiguous bins.
n_removed <- quast_output_plasmids_chromosomes %>% filter(non_ambiguous_length <= 0 & alignment_length != 0) %>% nrow()
n_fully_ambiguous <- quast_output_plasmids_chromosomes %>% filter(non_ambiguous_length <= 0 & alignment_length != 0 & fully_ambiguous==TRUE) %>% nrow()

print(paste0("Ambiguous alignments removed: ",n_removed))
print(paste0("Ambiguous alignments removed belonging to fully ambiguous bin: ",n_fully_ambiguous))
print(paste0("Percentage of alignments removed belonging to fully ambiguous bin: ",round(n_fully_ambiguous/n_removed*100,digits = 1),"%"))

#Remove fully ambiguous bin-replicon associations
quast_output_plasmids_chromosomes %<>% filter(non_ambiguous_length > 0 | alignment_length == 0)

```

```{r}
#Filter replicons classified as plasmids
quast_output_plasmids<-filter(quast_output_plasmids_chromosomes,classification=='plasmid')
#For cases in which the bin_name was not present, this means that the tool did not detected this plasmid. Assign that category.
quast_output_plasmids$bin_name<-ifelse(is.na(quast_output_plasmids$bin_name),"not_detected",as.character(quast_output_plasmids$bin_name))
#For this cases also assign an alignment length =0
quast_output_plasmids$alignment_length[is.na(quast_output_plasmids$alignment_length)]<-0
#assign not applicable to all the rest of NA
quast_output_plasmids[is.na(quast_output_plasmids)]<-'not_applicable'

```

### Calculating Recall
```{r}
quast_output_plasmids$recall_bp<-as.numeric(quast_output_plasmids$alignment_length)/as.numeric(quast_output_plasmids$replicon_length)
```

Now assign a 'not detected' category for recall=0 (to exclude this values from calculations
```{r}
quast_output_plasmids$recall_bp<-ifelse(quast_output_plasmids$recall_bp==0,'not_detected',as.numeric(quast_output_plasmids$recall_bp))
```

```{r}
#3. ------- Calculate precision
quast_output_plasmids$precision_bp<-as.numeric(quast_output_plasmids$alignment_length_precision)/as.numeric(quast_output_plasmids$bin_length)

#format the data and round the results
quast_output_plasmids$precision_bp<-as.numeric(quast_output_plasmids$precision_bp)
quast_output_plasmids$recall_bp<-as.numeric(quast_output_plasmids$recall_bp)
quast_output_plasmids$recall_bp<-round(quast_output_plasmids$recall_bp,3)
quast_output_plasmids$precision_bp<-round(quast_output_plasmids$precision_bp,3)


#4 --- Calculate f1-score
quast_output_plasmids$f1_score<-round(2*((quast_output_plasmids$precision_bp*quast_output_plasmids$recall_bp)/(quast_output_plasmids$precision_bp+quast_output_plasmids$recall_bp)),3)

#convert NA to 0 for scores
quast_output_plasmids[is.na(quast_output_plasmids)]<-0

```

Analyse how many bins are aligned to either plasmid, chromosome, plasmid and chromosome or not aligned at all.
```{r}
bin_analysis <- quast_output_plasmids_chromosomes %>% filter(!is.na(bin_name)) %>% group_by(software) %>% mutate(multiple_replicons=duplicated(bin_name) | duplicated(bin_name, fromLast = TRUE))
bin_analysis %<>% mutate(bin_type=ifelse(multiple_replicons==FALSE & classification=="chromosome","chromosome",NA))
bin_analysis %<>% mutate(bin_type=ifelse(multiple_replicons==FALSE & classification=="plasmid","one_plasmid",bin_type))
bin_analysis %<>% group_by(software,bin_name) %>% mutate(bin_type=ifelse(multiple_replicons==TRUE & n_distinct(classification)==1,"multiple_plasmids",bin_type))
bin_analysis %<>% group_by(software,bin_name) %>% mutate(bin_type=ifelse(multiple_replicons==TRUE & n_distinct(classification)!=1 & n()==2,"plasmid_and_chromosome",bin_type))
bin_analysis %<>% group_by(software,bin_name) %>% mutate(bin_type=ifelse(multiple_replicons==TRUE & n_distinct(classification)!=1 & n()!=2,"multiple_plasmids_and_chromosome",bin_type))
bin_analysis %<>% mutate(bin_type=ifelse(reference_id=="no_correct_alignments" | reference_id== "contig_length_below_1k","no_correct_alignments",bin_type))

bin_analysis %<>% ungroup()
count_bin_types <- bin_analysis %>% select(bin_name,software,bin_type) %>% unique()
```

```{r}
count_bin_types %<>% mutate(assembler=ifelse(grepl("uni",software),"Unicycler","bactofidia"))
count_bin_types %<>% mutate(method=ifelse(grepl("cleaned",software),"Reconstruction first","Standard MOB-suite"))
count_bin_types %<>% mutate(method=ifelse(grepl("filtered",software),"Selection first",method))
```


```{r}
level_order_methods <- c("Standard MOB-suite","Reconstruction first","Selection first")
level_order_types <- c("one_plasmid","multiple_plasmids","plasmid_and_chromosome","multiple_plasmids_and_chromosome","chromosome","no_correct_alignments")
ggplot(count_bin_types, aes(y=factor(method,level=rev(level_order_methods)),fill=factor(bin_type,level = rev(level_order_types))))+
  geom_bar(position = 'stack')+
  labs(x = "Number of bins",fill="Composition")+
  scale_fill_manual(values=rev(c("#75C24B","#47954D","#fc8e13","#ec6b3b","#e84141","#5c5d5f")),labels=rev(c("One plasmid","Multiple plasmids","Plasmid and chromosome","Multiple plasmids and chromosome","Chromosome","No correct alignments")),guide = guide_legend(reverse = TRUE))+
  labs(title=NULL)+
  theme(axis.title.y = element_blank())+
  facet_wrap(~assembler)
  
```

```{r}
ggsave("../results/figures/composition.png",width = 7,height = 3)
```

```{r}
level_order_types <- c("one_plasmid","multiple_plasmids","plasmid_and_chromosome","multiple_plasmids_and_chromosome","chromosome","no_correct_alignments")
ggplot(count_bin_types %>% filter(software %in% c('mob_uni','mob_uni_cleaned')), aes(y=factor(software,level=c('mob_uni_cleaned','mob_uni')),fill=factor(bin_type,level = rev(level_order_types))))+
  geom_bar(position = 'stack')+
  labs(x = "Number of bins",fill="Legend")+
  scale_fill_manual(values=rev(c("#75C24B","#47954D","#fc8e13","#ec6b3b","#e84141","#5c5d5f")),labels=rev(c("One plasmid","Multiple plasmids","Plasmid and chromosome","Multiple plasmids and chromosome","Chromosome","No correct alignments")),guide = guide_legend(reverse = TRUE))+
  labs(title="Bin composition per reconstruction method")+
  theme(axis.title.y = element_blank())+
  scale_y_discrete(labels=rev(c("MOB-suite","MOB-suite + plasmidEC")))
```

```{r}
ggsave("../results/figures/bin_analysis_pres.svg",width = 8,height = 3)
```

Composition of mixed bins
```{r}
mixed_bin_analysis <- bin_analysis %>% filter(classification=="chromosome") #& multiple_replicons==TRUE)
#mixed_bin_analysis %<>% left_join(multiple_ambiguous_alignments_final_substracion_data,by=c('strain','bin_name','reference_id','software'))
#replace NA values with 0 in the total_substraction_length (for cases in which there are no multiple alignments)
mixed_bin_analysis$total_substraction_length[is.na(mixed_bin_analysis$total_substraction_length)]<-0

#update alignment length
mixed_bin_analysis %<>% mutate(actual_alignment_length=alignment_length-total_substraction_length)
mixed_bin_analysis %<>% mutate(chromosomal_content=actual_alignment_length/bin_length*100)
```

```{r}
ggplot(mixed_bin_analysis,aes(x=chromosomal_content,fill=bin_type))+
  geom_histogram(binwidth = 5,breaks= seq(0,100,by=5))+
  scale_x_continuous(breaks= seq(0,100,by=20))+
  facet_wrap(~software)
```


Some plotting
```{r}
length_data <- replicon_data %>% filter(classification=="plasmid")
ggplot(length_data,aes(x=log(replicon_length,10)))+geom_histogram(binwidth = 0.05)
```

```{r}
plotting_data <- filter(quast_output_plasmids ,!software %in% c("not_applicable"))
level_order_softwares=c("MOB-suite (bactofidia)","MOB-suite (bactofidia, cleaned)","MOB-suite (bactofidia, plasmid contigs)","MOB-suite (unicycler)","MOB-suite (unicycler, cleaned)","MOB-suite (unicycler, plasmid contigs)")
plotting_data$software = str_wrap(plotting_data$software, width=18)
ggplot(plotting_data,aes(x=software,y=f1_score)) +geom_boxplot() +geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)+
labs(title = "F1 score per plasmid containing bin")
```

```{r}
ggsave("../results/figures/f1_bins.svg",width = 7, height = 4)
```

```{r}
plotting_data %<>% mutate(assembler=ifelse(grepl("uni",software),"Unicycler","bactofidia"))
plotting_data %<>% mutate(method=ifelse(grepl("cleaned",software),"Reconstruction first","Standard MOB-suite"))
plotting_data %<>% mutate(method=ifelse(grepl("filtered",software),"Selection first",method))
```

```{r}
bin_prec <- ggplot(plotting_data,aes(x=factor(method,level=level_order_methods),y=precision_bp)) +geom_boxplot(outlier.shape = NA) +geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)+
labs(title = "Precision per plasmid containing bin")+
  facet_wrap(~assembler)+
  labs(x = NULL, y = "Precision")+
  scale_x_discrete(labels = function(x) str_wrap(x, width = 10))
```

```{r}
ggsave("../results/figures/precision_bins.svg",width = 7, height = 4)
```

```{r}
bin_recall <- ggplot(plotting_data,aes(x=factor(method,level=level_order_methods),y=recall_bp)) +geom_boxplot(outlier.shape = NA) +geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)+
labs(title = "Recall per plasmid containing bin")+
  facet_wrap(~assembler)+
  labs(x = NULL, y = "Recall")+
  scale_x_discrete(labels = function(x) str_wrap(x, width = 10))
```

```{r}
ggsave("../results/figures/recall_bins.svg",width = 7, height = 4)
```

```{r}
ggplot(plotting_data,aes(x=recall_bp,y=precision_bp)) +
  geom_count(alpha=0.3) +
  scale_size_continuous(range = c(1.5,5)) +
  facet_wrap(~software)
```

### Calculate metrics per reference plasmid instead of per bin
```{r}
quast_output_reference <- quast_output_plasmids %>% group_by(strain,reference_id,replicon_length,software) %>% summarise(total_aligned=sum(alignment_length),total_aligned_precision=sum(alignment_length_precision), total_bins=sum(bin_length),n_bins=n())
quast_output_reference %<>% mutate(recall_bp=total_aligned/replicon_length,precision_bp=total_aligned_precision/total_bins,f1_score=2*(precision_bp*recall_bp)/(precision_bp+recall_bp))
```
```{r}
quast_output_reference %<>% mutate(assembler=ifelse(grepl("uni",software),"Unicycler","bactofidia"))
quast_output_reference %<>% mutate(method=ifelse(grepl("cleaned",software),"Reconstruction first","Standard MOB-suite"))
quast_output_reference %<>% mutate(method=ifelse(grepl("filtered",software),"Selection first",method))
```

```{r}
ref_prec <- ggplot(quast_output_reference,aes(x=factor(method,level=level_order_methods),y=precision_bp)) +geom_boxplot(outlier.shape = NA) +geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)+
labs(title = "Precision per reference plasmid")+
  facet_wrap(~assembler)+
  labs(x = NULL, y = "Precision")+
  scale_x_discrete(labels = function(x) str_wrap(x, width = 10))
```

```{r}
ref_recall <- ggplot(quast_output_reference,aes(x=factor(method,level=level_order_methods),y=recall_bp)) +geom_boxplot(outlier.shape = NA) +geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)+
labs(title = "Recall per reference plasmid")+
  facet_wrap(~assembler)+
  labs(x = NULL, y = "Recall")+
  scale_x_discrete(labels = function(x) str_wrap(x, width = 10))
```

Plot four plots together
```{r}
bin_recall+bin_prec+ref_recall+
  plot_layout(ncol = 1)+
  plot_annotation(tag_levels = 'A')
```
```{r}
ggsave("../results/figures/mobsuite_bins.svg",width = 10,height = 12)
```

```{r}
ggplot(quast_output_reference %>% filter(software %in% c('mob_uni','mob_uni_cleaned')),aes(x=software,y=recall_bp))+
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)+
  labs(title="Recall per reference plasmid",y="Recall")+
  theme(axis.title.x = element_blank())+
  scale_x_discrete(labels=c("MOB-suite","MOB-suite + plasmidEC"))
```
```{r}
ggsave("../results/figures/recall_reference_pres.svg",width = 6, height = 4)
```

```{r}
ggplot(quast_output_reference,aes(x=software,y=precision_bp))+
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)+
  labs(title="Recall per reference plasmid")
```
```{r}
ggsave("../results/figures/recall_reference.svg",width = 7, height = 4)
```

Find out whether a reference plasmid was predicted by the four methods, and whether it was split over bins.

```{r}
n_software <- 6
bin_number_plotting<-quast_output_plasmids %>% group_by(reference_id) %>% count(software)
not_detected<-quast_output_plasmids %>% filter(software=='not_applicable') %>% select(reference_id) %>% uncount(n_software)
not_detected %<>% mutate(software = rep(level_order_softwares,times=nrow(not_detected)/n_software)) %>% mutate(n = 0)
bin_number_plotting<-rbind(bin_number_plotting,not_detected)
bin_number_plotting %<>% spread(software,n) #%>% select(!not_applicable)
bin_number_plotting[is.na(bin_number_plotting)]<-0
bin_number_plotting %<>% gather('software','n',2:(n_software+1))
#bin_number_plotting %<>% mutate(correct=n==1)
```

Visualise by circle packing

```{r}
packing_data <- bin_number_plotting %>% group_by(software,n) %>% count() %>% rename(bins=n,count=nn)
```

```{r}
build_packing <- function(software_name) { 
  data <- packing_data %>% filter(software == software_name)
  packing <- circleRepelLayout(data$count,sizetype = "area")
  packing <- cbind(data,packing)
  return(packing)
  }
```

```{r}
softwares <- c("mob_uni","mob_bac","mob_uni_cleaned","mob_uni_filtered","mob_bac_cleaned","mob_bac_filtered")
dat.gg_all <- data.frame()

for (software in softwares){
  packing <- build_packing(software) %>% ungroup %>% select(x,y,radius,bins)
  #set custom x and y positions
  packing %<>% mutate(x=c(-11,5,-10,0,10,8),y=c(10,10,0,-5,-4,-8))
  dat.gg <- circleLayoutVertices(packing, sizetype = "radius", npoints=50, idcol = "bins")
  dat.gg %<>% mutate(software=software)
  name <- paste0("dat.gg_",software)
  assign(name,dat.gg)
  dat.gg_all %<>% rbind(.,dat.gg)
}
```


```{r}
ggplot(data = dat.gg_all, aes(x, y))+ 
  geom_polygon(aes(group = id,fill=as.factor(id))) +
  coord_equal()+
  facet_wrap(~software)
```
```{r}
packing_data %>% spread(bins,count)
```


```{r}
ggplot(packing_data, aes(x=count,y=as.factor(bins)))+
  geom_bar(stat = "identity")+
  facet_wrap(~software)
```


## ARG specific!

```{r}
AMR_replicons <- read.csv("../../binary_classifiers/results/arg_plasmids.csv", header = TRUE)
names(AMR_replicons) <- c("assembly_accession","replicon_name","atbr_count")
all_strains <- sub("_2021.*","\\1",all_contigs$contig_name)
AMR_replicons %<>% filter(assembly_accession %in% all_strains)
quast_output_AMR_plasmids <- inner_join(quast_output_plasmids,AMR_replicons,by=c("reference_id"="replicon_name"))
```

```{r}
plotting_data <- filter(quast_output_AMR_plasmids ,!software %in% c("not_applicable"))
level_order_softwares=c("MOB-suite (bactofidia)","MOB-suite (bactofidia, cleaned)","MOB-suite (bactofidia, plasmid contigs)","MOB-suite (unicycler)","MOB-suite (unicycler, cleaned)","MOB-suite (unicycler, plasmid contigs)")
plotting_data$software = str_wrap(plotting_data$software, width=18)
ggplot(plotting_data,aes(x=software,y=f1_score)) +geom_boxplot() +geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)+
labs(title = "F1 score per plasmid containing bin (AMR plasmids)")
```
```{r}
ggsave("../results/figures/F1_bins_AMR.svg",width = 7, height = 4)
```

```{r}
ggplot(plotting_data,aes(x=software,y=precision_bp)) +geom_boxplot(outlier.shape = NA) +geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)+
labs(title = "Precision per plasmid containing bin (AMR plasmids)")
```
```{r}
ggsave("../results/figures/precision_bins_AMR.svg",width = 7, height = 4)
```

```{r}
ggplot(plotting_data,aes(x=software,y=recall_bp)) +geom_boxplot(outlier.shape = NA) +geom_jitter(alpha=0.5,shape=16, position=position_jitter(0.2))+
labs(title = "Recall per plasmid containing bin (AMR plasmids)")
```

```{r}
ggsave("../results/figures/recall_bins_AMR.svg",width = 7, height = 4)
```

```{r}
ggplot(plotting_data,aes(x=recall_bp,y=precision_bp)) +
  geom_count(alpha=0.3) +
  scale_size_continuous(range = c(1.5,5)) +
  facet_wrap(~software)
```

### Calculate metrics per reference plasmid instead of per bin
```{r}
quast_output_AMR_plasmids_reference <- right_join(quast_output_plasmids,AMR_replicons,by=c("reference_id"="replicon_name"))
```

```{r}
quast_output_reference <- quast_output_AMR_plasmids_reference %>% group_by(strain,reference_id,replicon_length,software) %>% summarise(total_aligned=sum(alignment_length),total_aligned_precision=sum(alignment_length_precision), total_bins=sum(bin_length),n_bins=n())
quast_output_reference %<>% mutate(recall_bp=total_aligned/replicon_length,precision_bp=total_aligned_precision/total_bins,f1_score=2*(precision_bp*recall_bp)/(precision_bp+recall_bp))
```

```{r}
ggplot(quast_output_reference,aes(x=software,y=f1_score))+
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)
```

```{r}
ggplot(quast_output_reference,aes(x=software,y=recall_bp))+
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)
```

```{r}
ggplot(quast_output_reference,aes(x=software,y=precision_bp))+
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.5)
```

Find out whether a reference plasmid was predicted by the four methods, and whether it was split over bins.

```{r}
n_software <- 6
bin_number_plotting<-quast_output_AMR_plasmids_reference %>% group_by(reference_id) %>% count(software)
not_detected<-quast_output_plasmids %>% filter(software=='not_applicable') %>% select(reference_id) %>% uncount(n_software)
not_detected %<>% mutate(software = rep(level_order_softwares,times=nrow(not_detected)/n_software)) %>% mutate(n = 0)
bin_number_plotting<-rbind(bin_number_plotting,not_detected)
bin_number_plotting %<>% spread(software,n) #%>% select(!not_applicable)
bin_number_plotting[is.na(bin_number_plotting)]<-0
bin_number_plotting %<>% gather('software','n',2:(n_software+1))
#bin_number_plotting %<>% mutate(correct=n==1)
```

Visualise by circle packing

```{r}
packing_data <- bin_number_plotting %>% group_by(software,n) %>% count() %>% rename(bins=n,count=nn)
```

```{r}
build_packing <- function(software_name) { 
  data <- packing_data %>% filter(software == software_name)
  packing <- circleRepelLayout(data$count,sizetype = "area")
  packing <- cbind(data,packing)
  return(packing)
  }
```

```{r}
softwares <- c("mob_uni","mob_bac","mob_uni_cleaned","mob_uni_filtered","mob_bac_cleaned","mob_bac_filtered")
dat.gg_all <- data.frame()

for (software in softwares){
  packing <- build_packing(software) %>% ungroup %>% select(x,y,radius,bins)
  #set custom x and y positions
  packing %<>% mutate(x=c(-11,5,-10,0,10,8),y=c(10,10,0,-5,-4,-8))
  dat.gg <- circleLayoutVertices(packing, sizetype = "radius", npoints=50, idcol = "bins")
  dat.gg %<>% mutate(software=software)
  name <- paste0("dat.gg_",software)
  assign(name,dat.gg)
  dat.gg_all %<>% rbind(.,dat.gg)
}
```


```{r}
ggplot(data = dat.gg_all, aes(x, y))+ 
  geom_polygon(aes(group = id,fill=as.factor(id))) +
  coord_equal()+
  facet_wrap(~software)
```